service_capacity_modeling/models/org/netflix/elasticsearch.py

```diff
@@ -0,0 +1,471 @@
+from service_capacity_modeling.interface import DataShape
+from service_capacity_modeling.interface import Drive
+from service_capacity_modeling.interface import FixedInterval
+from service_capacity_modeling.interface import Instance
+from service_capacity_modeling.interface import Interval
+from service_capacity_modeling.interface import QueryPattern
+from service_capacity_modeling.interface import RegionContext
+from service_capacity_modeling.interface import Requirements
+from service_capacity_modeling.models import CapacityModel
+from service_capacity_modeling.models.common import compute_stateful_zone
+from service_capacity_modeling.models.common import simple_network_mbps
+from service_capacity_modeling.models.common import sqrt_staffed_cores
+from service_capacity_modeling.models.common import working_set_from_drive_and_slo
+from service_capacity_modeling.stats import dist_for_interval
+
+
+logger = logging.getLogger(__name__)
+
+
+def _target_rf(desires: CapacityDesires, user_copies: Optional[int]) -> int:
+    if user_copies is not None:
+        assert user_copies > 1
+        return user_copies
+
+    # Due to the relaxed durability and consistency requirements we can
+    # run with RF=2
+    consistency = desires.query_pattern.access_consistency.same_region
+    if (
+        desires.data_shape.durability_slo_order.mid < 1000
+        and consistency is not None
+        and consistency.target_consistency != AccessConsistency.read_your_writes
+    ):
+        return 2
+    return 3
+
+
+# Looks like Elasticsearch (Lucene) uses a tiered merge strategy of 10
+# segments of 512 megs per
+# https://lucene.apache.org/core/8_1_0/core/org/apache/lucene/index/TieredMergePolicy.html#setSegmentsPerTier(double)
+# (FIXME) Verify what elastic merge actually does
+def _es_io_per_read(node_size_gib, segment_size_mb=512):
+    gb = node_size_gib * 1024
+    segments = max(1, gb // segment_size_mb)
+    # 10 segments per tier, plus 1 for L0 (avg)
+    levels = 1 + int(math.ceil(math.log(segments, 10)))
+    return levels
+
+
+def _estimate_elasticsearch_requirement(
+    instance: Instance,
+    desires: CapacityDesires,
+    working_set: float,
+    reads_per_second: float,
+    max_rps_to_disk: int,
+    zones_per_region: int = 3,
+    copies_per_region: int = 3,
+) -> CapacityRequirement:
+    """Estimate the capacity required for one zone given a regional desire
+
+    The input desires should be the **regional** desire, and this function will
+    return the zonal capacity requirement
+    """
+    # Keep half of the cores free for background work (merging mostly)
+    needed_cores = math.ceil(sqrt_staffed_cores(desires) * 1.5)
+    # Keep half of the bandwidth available for backup
+    needed_network_mbps = simple_network_mbps(desires) * 2
+
+    needed_disk = math.ceil(
+        (1.0 / desires.data_shape.estimated_compression_ratio.mid)
+        * desires.data_shape.estimated_state_size_gib.mid
+        * copies_per_region,
+    )
```

Comment:
Shouldn't needed_cores, needed_network_mbps, and needed_disk be divided by zones_per_region? It seems like these are regional calculations though Elasticsearch shards are distributed across AZ's